lib/common/rules.c

```diff
@@ -1293,3 +1293,133 @@ pcmk__evaluate_op_expression(const xmlNode *op_expression,
               id, name, pcmk__readable_interval(rule_input->op_interval_ms));
     return pcmk_rc_ok;
 }
+
+/*!
+ * \internal
+ * \brief Evaluate a single rule expression, including any subexpressions
+ *
+ * \param[in,out] expr         XML containing a rule expression
+ * \param[in]     rule_input   Values used to evaluate rule criteria
+ * \param[out]    next_change  If not NULL, set to when evaluation will change
+ *
+ * \return Standard Pacemaker return code (\c pcmk_rc_ok if the expression
+ *         passes, some other value if it does not)
+ */
+int
+pcmk__evaluate_condition(xmlNode *expr, const pcmk_rule_input_t *rule_input,
+                         crm_time_t *next_change)
+{
+    int rc = pcmk_rc_ok;
+    const char *uname = NULL;
+
+    switch (pcmk__expression_type(expr)) {
+        case pcmk__subexpr_rule:
+            rc = pcmk_evaluate_rule(expr, rule_input, next_change);
+            break;
+
+        case pcmk__subexpr_attribute:
+        case pcmk__subexpr_location:
+            /* these expressions can never succeed if there is
+             * no node to compare with
+             */
+            if (rule_input->node_attrs != NULL) {
+                rc = pcmk__evaluate_attr_expression(expr, rule_input);
```

Comment:
This and the other helper functions should be updated to use "condition" at some point

Reply:
They're correct -- a condition is a subrule, or an expression of any type. They evaluate specific expression types.

Reply:
Oh okay, I thought we were moving from "expression" to "condition" in general but I see what you mean.